File: ScalaComplex/src/complex/AbsComplex.scala
package complex

abstract class AbsComplex extends Complex {
  override def +(c: Complex): Complex = this.create(this.real + c.real, this.imag + c.imag)
  override def *(c: Complex): Complex = this.create(this.radius * c.radius, this.azimuth + c.azimuth)

  override def equals(o: Any): Boolean = {
    if (!o.isInstanceOf[Complex]) return false
    val c: Complex = o.asInstanceOf[Complex]
    Comparison.isEqual(this.real, c.real) && Comparison.isEqual(this.imag, c.imag)
  }

  override val toString: String = "(" + this.real + ", " + this.real + ")"
}

File: TP1/ComplexJava/src/complex/AbsComplex.java
package complex;

public abstract class AbsComplex implements Complex {
	public Complex add(Complex c){
		return this.create(this.getReal() + c.getReal(),
				this.getImag() + c.getImag());
	}
	public Complex multiply(Complex c){
		return this.create(this.getRadius() * c.getRadius(),
				this.getAzimuth().add(c.getAzimuth()));
	}

	public boolean equals(Object o){
		if(!(o instanceof Complex)) return false;
		Complex c = (Complex)o;
		return Comparison.isEqual(this.getReal(), c.getReal()) 
				&& Comparison.isEqual(this.getImag(), c.getImag());
	} 
	public String toString(){
		return "(" + this.getReal() + ", " + this.getImag() + ")";
	}
}